BEST_PRACTICES.md

```diff
@@ -0,0 +1,146 @@
+
+_**Q6. I want to perform operations on Satellite/Capsule/ContentHost like execute CLI commands, read conf files etc, where should I add a method for that?**_
+
+  _Answer:_ Add into existing mixin classes in mixin modules in `robottelo/host_helpers/*_mixins.py` or create a new targetted mixin class and inherit than in host classes of `robottelo/hosts.py` module
+
+_**Q7. I have new CLI or UI or API entity or operation to implement, should it be in factory object or endpoint object.**_
+
+  _Answer:_ New entity/subcommand/operations added in satellite product should be added to be accessed by endpoint onjects. e.g `target_sat.api.ComponentNew()`. The factory objects are just for adding helpers.
+
+_**Q8. When should I prefer helper function as a host object method over utils function ?**_
+
+  _Answer:_ When the new function is dependent on existing host methods or attributes then its good implement such helper function in host classes as it makes to access those host methods and attributes easier.
+
+_**Q9. What helper functions goes into `robottelo/utils/__init__.py` module?**_
+  _Answer:_ If the helper does not fit into any of the existing utils modules or subject of helper(s) is not big enough to create new , then add in `robottelo/utils/__init__.py` as a standalone helper.
+
+_**Q10. I have a helper function that applies to all Satellite, Capsule, ContentHost classes. Which class should I choose to add it?**_
+
+  _Answer:_ The recommeded class in such cases is always the highest parent class. Here ContentHost class is the highest parent class where this function should be added as a class method or a mixin method.
+
+_**Q11. I have some upgrade scenario helpers to implement, where should I add them?**_
+
+  _Answer:_ Upgrade scenarios are not special in this case. All helpers/function except framework should be same as being used in foreman tests and hence all rules are applied.
+
+_**Q12. Where does a non-reusable helper function should reside?**_
+
+  _Answer:_ The preffered place is the test module where its being used.
+
+_**Q13. I need to extend the functionality of third party library methods/objects, should I do it in robottelo?**_
+
+  _Answer:_ Its recommended to extend the functionality of third library methods in that library itself if its being maintained by SatelliteQE like airgun,nailgun, manifester or is active community like widgetastic, wrapanapi. Else extend that appropriately in utils, host methods or fixtures.
+
+_**Q14. What if I see two helper methods/functions almost doing equal operations with a little diff ?**_
+
+  _Answer:_ See if this could be merged in one function with optional parameters else leave them separated. E.g provisioning functions using API calls with minimum difference and being used in API, UI and CLI tests then merge them. But if two provisioing helpers one for CLI and API tests and tests demands it then good to keep it separate.
```

Comment:
```suggestion
  _Answer:_ See if this could be merged in one function with optional parameters else leave them separated. E.g provisioning functions using API calls with minimum difference and being used in API, UI and CLI tests then merge them. But if two provisioning helpers one for CLI and API tests and tests demands it then good to keep it separate.
```